### ISSUE TITLE
REP STOSW instruction is disassembled as REP STOSD in 32-bit mode

### ISSUE THREAD

**Issue by 10110111** [aquynh/capstone#917]
This is a sequel to #452. The problem is the same as there, but for 32-bit instructions. Example error:
```
Intel32: 0x9123ffe1:       67 66 f3 ab     rep stosw word ptr es:[di], ax
Expected string is: rep stosw word ptr es:[di], ax
Capstone returned : rep stosd dword ptr es:[di], eax
```
To see related errors, clone my repository of https://github.com/10110111/capstone-experiments and say there
```
make tests && ./tests string-insns-test.txt
```
You'll get 36 test failures, all of which are related to this issue.

**Comment by aquynh** [comment]
i have a local fix, which passed everything, except the below code:

```
Intel32: 0x9123ffe1:       66 f3 6f        rep outsw dx, dword ptr [esi]     ; reg16, mem16
string-insns-test.txt:199: note: expected string is: rep outsw dx, dword ptr [esi]
string-insns-test.txt:199: note: capstone returned : rep outsw dx, word ptr [esi]
```

this is "outsw" instruction, so are you sure your expected result is correct?

**Comment by hlide** [comment]
it should be: 16-bit `rep outsw dx, word ptr [esi]` when being in default 32/64-bit mode or 32-bit `rep outsd dx, dword ptr [esi]` when being in default 16-bit mode. This expected string for this test is clearly wrong.

**Comment by aquynh** [comment]
@hlide,  thanks for chiming in. here is the current output with https://github.com/aquynh/capstone/tree/test.

```
$ cstool x16 "66 f3 6f"
0  66f36f                            rep outsd	dx, dword ptr [si]

$ cstool x32 "66 f3 6f"
0  66f36f                            rep outsw	dx, word ptr [esi]

$ cstool x64 "66 f3 6f"
0  66f36f                            rep outsw	dx, word ptr [rsi]
```

**Comment by 10110111** [comment]
Yeah, the expected string is wrong indeed. Seems I haven't taken something into account in my generator.

**Comment by aquynh** [comment]
your repo is very useful. do you want to make a pull req to put it under https://github.com/aquynh/capstone/tree/next/suite?

**Comment by 10110111** [comment]
Hmm, currently it's somewhat disorganized. Are you sure you want to merge it in its current form? Or do you plan to reshape it yourself after merge?

**Comment by aquynh** [comment]
what is wrong with the current code?

**Comment by 10110111** [comment]
Mainly its non-uniformity:

* `*.c` files just print what has been disassembled, without checking for correctness
* `tests.cpp` does the tests based on `*.txt` input (the one I originally suggested in this report)
* `string-insns-test-gen.cpp` generates test data for string instructions to be consumed by `tests.cpp` binary
* `*.txt` are manually-crafted test files (thus the mistake with `dword ptr` discussed above was a typo)

If you're OK with this, I'll do a PR.

**Comment by aquynh** [comment]
looks like all your testcases are already in https://github.com/aquynh/capstone/blob/next/suite/regress.py ?

**Comment by 10110111** [comment]
Seems not all. In particular, I've found neither `rep outsw dx, dword ptr [esi]` (with the typo discussed above) nor `\x66\xf3\x6f` in that file.

**Comment by aquynh** [comment]
do you mind adding the missing instructions there?

we can maintain all the regression tests there from now on.

**Comment by aquynh** [comment]
btw, there are some bugs in that file, in which the "expected output" are wrong, and needs to be fixed.

**Comment by 10110111** [comment]
I'll try doing it ASAP.

**Comment by aquynh** [comment]
awesome! while you are at it, if you see any bugs in that file, please fix them, thanks.

**Comment by 10110111** [comment]
> btw, there are some bugs in that file, in which the "expected output" are wrong, and needs to be fixed.
Could you cite some of them? I'm currently checking the `next` branch, and first several errors are indeed errors in capstone:

```
X86-32bit intel: 0x66 0xf2 0x6d =  repne insd dword ptr es:[edi], dx
         --> ERROR: expected output = repne insw word ptr es:[edi], dx
X86-32bit intel: 0x66 0xf2 0x6f =  repne outsd dx, dword ptr [esi]
         --> ERROR: expected output = repne outsw dx, word ptr [esi]
X86-32bit intel: 0x66 0xf2 0xa5 =  repne movsd dword ptr es:[edi], dword ptr [esi]
         --> ERROR: expected output = repne movsw word ptr es:[edi], word ptr [esi]
```

**Comment by aquynh** [comment]
Yes, i also saw these, as they stand out. There maybe others, but i dont remember all of them.
